nav2_recoveries/plugins/assisted_teleop.cpp

```diff
@@ -0,0 +1,300 @@
+#include <string>
+
+#include "assisted_teleop.hpp"
+#include "nav2_msgs/action/assisted_teleop.hpp"
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_recoveries
+{
+using namespace std::chrono_literals; //NOLINT
+void
+AssistedTeleop::cleanup()
+{
+  action_server_.reset();
+  vel_pub_.reset();
+  vel_sub_.reset();
+  costmap_sub_.reset();
+}
+
+void
+AssistedTeleop::activate()
+{
+  RCLCPP_INFO(logger_, "Activating %s", recovery_name_.c_str());
+
+  vel_pub_->on_activate();
+  action_server_->activate();
+  enabled_ = true;
+}
+
+void
+AssistedTeleop::deactivate()
+{
+  vel_pub_->on_deactivate();
+  action_server_->deactivate();
+  enabled_ = false;
+}
+
+void
+AssistedTeleop::configure(
+  const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+  const std::string & name, std::shared_ptr<tf2_ros::Buffer> tf,
+  std::shared_ptr<nav2_costmap_2d::CostmapTopicCollisionChecker> collision_checker)
+{
+  node_ = parent;
+  auto node = node_.lock();
+
+  logger_ = node->get_logger();
+
+  RCLCPP_INFO(logger_, "Configuring %s", name.c_str());
+
+  recovery_name_ = name;
+  tf_ = tf;
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "projection_time", rclcpp::ParameterValue(1.0));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "linear_velocity_threshold_", rclcpp::ParameterValue(0.06));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "cmd_vel_topic", rclcpp::ParameterValue(std::string("cmd_vel")));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "input_vel_topic", rclcpp::ParameterValue(std::string("cmd_vel_input")));
+
+  node->get_parameter("global_frame", global_frame_);
+  node->get_parameter("robot_base_frame", robot_base_frame_);
+  node->get_parameter("transform_tolerance", transform_tolerance_);
+  node->get_parameter("projection_time", projection_time_);
+  node->get_parameter("linear_velocity_threshold_", linear_velocity_threshold_);
+  node->get_parameter("cmd_vel_topic", cmd_vel_topic_);
+  node->get_parameter("input_vel_topic", input_vel_topic_);
+
+  vel_sub_ = node->create_subscription<geometry_msgs::msg::Twist>(
+    input_vel_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::vel_callback,
+      this, std::placeholders::_1));
+
+  vel_pub_ = node->create_publisher<geometry_msgs::msg::Twist>(cmd_vel_topic_, 1);
+
+  costmap_sub_ = std::make_unique<nav2_costmap_2d::CostmapSubscriber>(
+    node_, "local_costmap/costmap_raw");
+
+  action_server_ = std::make_shared<ActionServer>(
+    node, recovery_name_,
+    std::bind(&AssistedTeleop::execute, this));
+
+  collision_checker_ = collision_checker;
+}
+
+void
+AssistedTeleop::vel_callback(const geometry_msgs::msg::Twist::SharedPtr msg)
+{
+  speed_x = msg->linear.x;
+  speed_y = msg->linear.y;
+  angular_vel_ = msg->angular.z;
+  if (go) {
+    if (!checkCollision()) {
+      RCLCPP_INFO(logger_, "Reducing velocity by a factor of %.2f", scaling_factor);
+    }
+    moveRobot();
```

Comment:
Pass values to this, don't do it implicitly with member variables being changed, that's bad design